### PR TITLE
add decompiler code, tests, and decompiler library files for v2

### PR DESCRIPTION
**Pull request by nwalsh1995**
As of right now there is no frontend to this code, and it is being activated by unit tests.

The testing runs each decompiler separately, and also runs them async. 

**Design decisions:**

1. I've taken to using `raise SystemExit("Exit Message")` if something is wrong instead of the old `sys.exit` approach QARK took in V1.
2. File structure, breaking each piece into its own package. Not entirely sold on this but it organizes the repo a bit better than having all the modules in the top level.
3. I am open to removing some of the helper methods in `decompiler.py` and putting them in a module like `qark/qark/decompiler/helpers.py`
4. Defaulting the build directory to the same directory as the APK. We can create our own directory as a subdirectory here which might make it cleaner, but I am unsure how much we should do.

Current testing gives me about 80% code coverage.

### CODE REVIEW

**Comment by neilatli** [comment]
> We can create our own directory as a subdirectory here which might make it cleaner, but I am unsure how much we should do.

I think this is the right approach, it makes it much easier for someone to clean up the artifacts from the run and run it again with different options.

File: qark/decompiler/decompiler.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+DECOMPILER_COMMANDS = {"JDCORE": "java -jar {path_to_decompiler} {jar} {build_directory}/jdcore",
+                       "PROCYON": "java -jar {path_to_decompiler} {jar} -o {build_directory}/procyon",
+                       "CFR": "java -jar {path_to_decompiler} {jar} --outputdir {build_directory}/cfr"}
+
+
+class Decompiler(object):
+    """This class handles unpacking and decompiling APKs into Java source code.
+
+    New compilers can be added by adding the command to run the compiler to the `DECOMPILER_COMMANDS` dictionary and
+    adding them into the `Decompiler.decompilers` dictionary with the path to the decompiler on the system.
```
Comment (neilatli) [inline qark/decompiler/decompiler.py]: > +    New compilers can be added by adding the command to run the compiler to the `DECOMPILER_COMMANDS` dictionary and
> +    adding them into the `Decompiler.decompilers` dictionary with the path to the decompiler on the system.

These seems inextricably linked and should just go in a small helper class. Something along the lines of:
```
class ExternalDecompiler(object):
    def __init__(self, name, local_path, command=None):
        self.name = name.lower()
        self.local_path = local_path
        self.command = command or 'java -jar {local_path} {build_directory}/{name}'.format(local_path=self.local_path, build_directory='{build_directory}', name=self.name)

jd_core = ExternalDecompiler('JDCORE', '/path/to/jdcore')

list_of_decompilers = [jd_core]

decompiler = Decompiler(decompilers=list_of_decompilers)
```

Or you could make ExternalDecompiler an ABC and subclass it for each decompiler. Either way, I think a helper class is the way to go.

Comment (nwalsh1995) [reply qark/decompiler/decompiler.py]: This is fixed in the newest commit. ExternalDecompiler is made into an ABC.

File: lib/apktool/darwin/apktool [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (nwalsh1995) [reply lib/apktool/darwin/apktool]: Fixed in the newest commit. We will automatically download cfr, procyon, apktool, and dex2jar. jd-core is submoduled but APKTool is not for size issues.

**Pull request merged**